### PR TITLE
[ENH] Add Yeo Johnson Transform

### LINKED ISSUE #1475: [ENH] Implement Yeo-Johnson Series Transformer
### Describe the feature or idea you want to propose

a long time ago I experimented with the Yeo-Johnson regressor transform.

https://ieeexplore.ieee.org/document/1716845

Yeo-Johnson is an extension of Box-Cox that allows xero and negative regressor values. We have an implementation of Box Cox, it would be nice to include Yeo-Johnson 

### Describe your proposed solution

either extend Box Cox once it is moved into series tranformers, or implement a new BaseSeriesTransformer

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alexbanwell1**
#### Reference Issues/PRs

Fixes #1475

#### What does this implement/fix? Explain your changes.

I have added the basic YeoJohnson Transform to the series transform section. This uses the scipy implementation to compute the transform, but implements the aeon interface on top of this.

#### Does your contribution introduce a new dependency? If yes, which one?

No new dependencies

#### Any other comments?

N/A

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

### CODE REVIEW

**Comment by alexbanwell1** [comment]
@all-contributors please add @alexbanwell1 for code, design and documentation

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
thanks for this, but given its so simple it might be nice to have our own implementation, then test it is same as scipy? That way its easy to make multivariate, just do each channel separately.

File: aeon/transformations/series/_yeojohnson.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        "univariate-only": True,
+        "requires_y": False,
+        "capability:inverse_transform": False,
+    }
+
+    def __init__(self, bounds=None):
+        self.bounds = bounds
+        super().__init__()
+
+    def _fit(self, X, y=None):
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/transformations/series/_yeojohnson.py:97]: actually ignore that

File: aeon/transformations/series/tests/test_yeojohnson.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+"""Tests for YeoJohnsonTransformer."""
+
+__maintainer__ = [Alex Banwell]
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/transformations/series/tests/test_yeojohnson.py]: Done

**Comment by TonyBagnall** [comment]
ah failed pre-commit, I got the tag name wrong :) Good to install pre-commit locally though

https://www.aeon-toolkit.org/en/stable/developer_guide/dev_installation.html

pip install --editable .[dev]

you should not actually need these capability tags, as they are the same as in BaseSeriesEstimator

**Comment by alexbanwell1** [comment]
I have implemented the YeoJohnson Transform natively in the code, and tested it against the scipy implementation as requested!

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
good stuff, very minor changes now

File: aeon/transformations/series/_yeojohnson.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    ----------
+    bounds : tuple
+        Lower and upper bounds used to restrict the feasible range
+        when solving for the value of lambda.
+    lambda_ : float
+        The Yeo-Johnson lambda parameter. If not supplied, it is solved for based
+        on the data provided in `fit`.
+
+    Attributes
+    ----------
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/transformations/series/_yeojohnson.py]: Changed lambda_ to lmbda

File: aeon/transformations/series/_yeojohnson.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        y : ignored argument for interface compatibility
+            Additional data, e.g., labels for transformation
+
+        Returns
+        -------
+        self: a fitted instance of the estimator
+        """
+        if self.lambda_ is None:
+            X = X.flatten()
+            self.lambda_ = yeojohnson_normmax(X, brack=self.bounds)
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/transformations/series/_yeojohnson.py]: Done!

File: aeon/transformations/series/_yeojohnson.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        Returns
+        -------
+        Xt : 2D np.ndarray
+            transformed version of X
+        """
+        return _yeo_johnson_transform(X, self.lambda_)
+
+
+@njit
+def _yeo_johnson_transform(X, lambda_):
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/transformations/series/_yeojohnson.py]: Done! Parameter is now called lmbda

**Review by TonyBagnall — APPROVED** [review]
good job, thanks

**Comment by TonyBagnall** [comment]
@all-contributors add @alexbanwell1 as a contributor for code, design, doc.

**Pull request merged**